### PR TITLE
MRG: Initialize tfr average

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
See #2523

Todo: test

### CODE REVIEW

File: mne/time_frequency/tfr.py [hunk 1]
```diff
@@ -1374,3 +1375,112 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+    Selects the subset of good channels across all datasets.
+
+    The grand_average.nave attribute will be equal the number
+    of datasets used to calculate the grand average.
+
+    Note: Grand average TFR shall not be used for source localization.
+
+    Parameters
+    ----------
+    all_tfr : list of AllTFR data
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: AllTFR -> AverageTFR

File: mne/time_frequency/tfr.py [hunk 2]
```diff
@@ -1374,3 +1375,112 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+        The TFR datasets.
+    drop_bads : bool
+        If True, bad MEG and EEG channels are dropped.
+
+    Returns
+    -------
+    grand_average : AllTFR
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: Same here

File: mne/time_frequency/tfr.py [hunk 3]
```diff
@@ -1374,3 +1375,112 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+    grand_average = combine_tfr(all_tfr)
+    # change the grand_average.nave to the number of Evokeds
+    grand_average.nave = len(all_tfr)
+    # change comment field
+    grand_average.comment = "Grand average (n = %d)" % grand_average.nave
+    return grand_average
+
+
+def combine_tfr(all_tfr, weights='nave'):
+    """Merge TFR data
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: And here

File: mne/time_frequency/tfr.py [hunk 4]
```diff
@@ -1374,3 +1375,112 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+
+    Data should have the same channels and the same time instants.
+    Subtraction can be performed by passing negative weights (e.g., [1, -1]).
+
+    Parameters
+    ----------
+    all_tfr : list of TFR
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: here

File: mne/time_frequency/tfr.py [hunk 5]
```diff
@@ -1374,3 +1375,112 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+        The tfr datasets.
+    weights : list of float | str
+        The weights to apply to the data of each TFR instance.
+        Can also be ``'nave'`` to weight according to tfr.nave,
+        or ``"equal"`` to use equal weighting (each weighted as ``1/N``).
+
+    Returns
+    -------
+    tfr : AllTFR
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: here :)

File: mne/time_frequency/tfr.py [hunk 6]
```diff
@@ -1374,3 +1375,112 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+            weights = np.array([e.nave for e in all_tfr], float)
+            weights /= weights.sum()
+        else:  # == 'equal'
+            weights = [1. / len(all_tfr)] * len(all_tfr)
+    weights = np.array(weights, float)
+    if weights.ndim != 1 or weights.size != len(all_tfr):
+        raise ValueError('Weights must be the same size as all_tfr')
+
+    ch_names = tfr.ch_names
+    for e in all_tfr[1:]:
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: e -> this_tfr

File: mne/time_frequency/tfr.py [hunk 7]
```diff
@@ -1374,3 +1375,112 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+                       % (tfr, e))
+
+    # use union of bad channels
+    bads = list(set(tfr.info['bads']).union(*(ev.info['bads']
+                                              for ev in all_tfr[1:])))
+    tfr.info['bads'] = bads
+
+    tfr.data = sum(w * e.data for w, e in zip(weights, all_tfr))
+    tfr.nave = max(int(1. / sum(w ** 2 / e.nave
+                                for w, e in zip(weights, all_tfr))), 1)
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: I'm not sure what you're aiming at here, but shouldn't this be just ``tfr.nave = len(all_tfr)``.

Comment (agramfort) [inline mne/time_frequency/tfr.py]: >
> I'm not sure what you're aiming at here, but shouldn't this be just tfr.nave
> = len(all_tfr).
>
same behavior as for evoked. We should be consistent.

Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: Yeah. In this case it's actually changed back to ``len(all_tfr)`` at line 1425.
In my opinion the ``combine_tfr`` should be made private and ``weights`` param exposed to the user in ``grand_average``. Same goes for evoked. Is it too late to change them?

Comment (agramfort) [inline mne/time_frequency/tfr.py]: combine_tfr or combine_evoked allows you to contrast data too using
negative weights.

Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: Okay, We had a chat with Alex offline. The ``grand_average`` should use ``equal`` as the method for averaging just like with evoked. Add that on line 1423.

Comment (jona-sassenhagen) [inline mne/time_frequency/tfr.py]: ok

**Comment by jona-sassenhagen** [comment]
@jaeilepp review? (Travis is fine but appveyor hangs, not due to me I think?)

Note, I haven't implemented subject subset averaging for now.

File: mne/time_frequency/tfr.py [hunk 8]
```diff
@@ -1374,3 +1376,112 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+    -------
+    tfr : AverageTFR
+        The new TFR data.
+
+    Notes
+    -----
+    .. versionadded:: 0.10.0
+    """
+    tfr = all_tfr[0].copy()
+    if isinstance(weights, string_types):
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: I think you can use ``isinstance(weights, str)`` here to avoid the import.

File: mne/time_frequency/tfr.py [hunk 9]
```diff
@@ -1374,3 +1376,112 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+        if weights not in ('nave', 'equal'):
+            raise ValueError('Weights must be a list of float, or "nave" or '
+                             '"equal"')
+        if weights == 'nave':
+            weights = np.array([tfr_.nave for tfr_ in all_tfr], float)
+            weights /= weights.sum()
+        else:  # == 'equal'
+            weights = [1. / len(all_tfr)] * len(all_tfr)
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: A nitpick, but I think 
```
if weights == 'nave':
...
elif weights == 'equal'
...
else:
raise ValueError ...
```
is nicer.

File: mne/time_frequency/tfr.py (hunk 8 above)
Comment (Eric89GXL) [inline mne/time_frequency/tfr.py]: We want to use `string_types` as much as possible for maximum compatibilty with ASCII/Unicode strings on Python2. `isinstance(u'foo', str)` will fail on Python2, for example.

Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: Oh, okay then. Good to know.

File: mne/time_frequency/tfr.py [hunk 10]
```diff
@@ -1374,3 +1376,112 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+    # Copy channels to leave the original evoked datasets intact.
+    all_tfr = [tfr_.copy() for tfr_ in all_tfr]
+
+    # Interpolates if necessary
+    if drop_bads:
+        for tfr in all_tfr:
+            tfr.drop_channels(tfr.info['bads'], copy=False)
+    else:
+        pass
+        # implement good channel subset here somehow??
```
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: I think you can remove this? Seems to work fine as it is.

**Comment by jaeilepp** [comment]
Otherwise LGTM.

File: mne/time_frequency/tfr.py (hunk 10 above)
Comment (jona-sassenhagen) [inline mne/time_frequency/tfr.py]: That's my placeholder for doing channel-wise subject subset averaging ... I think I should raise a NotImplementedError here?

File: mne/time_frequency/tfr.py (hunk 9 above)
Comment (jona-sassenhagen) [inline mne/time_frequency/tfr.py]: This is how it is in evoked.grand_average.

Comment (jona-sassenhagen) [inline mne/time_frequency/tfr.py]: I'd keep it like this for now for the sake of consistency.

File: mne/time_frequency/tfr.py (hunk 10 above)
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: I don't quite understand what you mean...
In my opinion, you can just leave this out. If ``drop_bad=False`` then you can just continue without dropping the bads. You can then just implement it later, unless you want to include it in this PR.

File: mne/time_frequency/tfr.py (hunk 9 above)
Comment (jaeilepp) [inline mne/time_frequency/tfr.py]: Yeah, it's not a biggie.

**Comment by jona-sassenhagen** [comment]
Thanks @jaeilepp 
Merge if you feel like it?

**Comment by jaeilepp** [comment]
Does someone else want to take a look?

**Comment by Eric89GXL** [comment]
Nah I trust you (and @jona-sassenhagen and the CIs) @jaeilepp :)

**Comment by jona-sassenhagen** [comment]
Don't trust me @Eric89GXL , I'm terrible :) And the CIs are only as good as the tests I've written.

**Comment by jona-sassenhagen** [comment]
But I think this should be fine.

**Comment by jaeilepp** [comment]
Yeah, actually there's no test for the data itself. Especially for the ``nave`` case it would be nice to have a simple test...

File: mne/time_frequency/tests/test_tfr.py [hunk 11]
```diff
@@ -127,6 +128,23 @@ def test_time_frequency():
...
     assert_equal(power_pick.data.shape, power_drop.data.shape)
 
+    # grand average
+    picks = pick_types(raw.info, meg='grad', eeg=False,
+                       stim=False, include=include,
+                       exclude=exclude)[:3]
+    epochs = Epochs(raw, events, event_id, tmin, tmax, picks=picks,
+                    baseline=(None, 0))
+    power, itc = tfr_morlet(epochs, freqs=freqs, n_cycles=n_cycles,
+                            use_fft=True, return_itc=True)
```
Comment (agramfort) [inline mne/time_frequency/tests/test_tfr.py]: why not reusing the power, itc from above? our test suite is already too long

File: mne/time_frequency/tfr.py [hunk 12]
```diff
@@ -1374,3 +1376,110 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
                                method='mutlitaper-itc'))
     return out
+
+
+def grand_average(all_tfr, drop_bads=True):
+    """Make grand average of a list TFR data
+
+    Selects the subset of good channels across all datasets.
+
+    The grand_average.nave attribute will be equal the number
```
Comment (agramfort) [inline mne/time_frequency/tfr.py]: equal to

File: mne/time_frequency/tfr.py [hunk 13]
```diff
@@ -1374,3 +1376,110 @@ def tfr_multitaper(inst, freqs, n_cycles, time_bandwidth=4.0,
...
+    grand_average : AverageTFR
+        The grand average data.
+
+    Notes
+    -----
+    .. versionadded:: 0.10.0
+    """
+    # check if all elements in the given list are evoked data
+    if not all(isinstance(tfr_, AverageTFR) for tfr_ in all_tfr):
+        raise ValueError("Not all the elements in list are AllTFR data")
```
Comment (agramfort) [inline mne/time_frequency/tfr.py]: AllTFR ?

**Comment by agramfort** [comment]
besides LGTM but I am wondering if there is a way to have a single grand_average function that can take Evoked or AverageTFR. It would avoid code duplication and avoid a new function.

**Comment by jona-sassenhagen** [comment]
@agramfort I too think so - something like mne.grand_average, which can take evoked, tfr, and possibly also GAT and other objects. EEGLAB has a function for averaging ERPimages.

The question is, implement as-is or refactor later? I don't think implementing as-is would be a waste, or make refactoring harder, right?

**Comment by jona-sassenhagen** [comment]
@jaeilepp you got an idea for how to test data itself?

**Comment by agramfort** [comment]
a mne.grand_average would be great

**Comment by jona-sassenhagen** [comment]
Now, or later? I think it would take me a bit of time to do that myself. And Eric wants me to do the ROI plot you don't like :)

**Comment by agramfort** [comment]
after the release. We'll try to release tomorrow. Last big question is do
we merge maxwell_filter before or after?

**Comment by jaeilepp** [comment]
> you got an idea for how to test data itself?

Just set the tfr data to some simple values (perhaps just one or two channels) and call ``combine_tfr`` with it. Then assert the returned data. It is just for making sure it is doing what it's supposed to.

**Comment by jona-sassenhagen** [comment]
I'm closing this one and reopening soon with a unified function, as discussed here #2543
Thanks for the help @jaeilepp 

Question, what should grand_average sit under? mne.grand_average?

**Comment by Eric89GXL** [comment]
Yeah, it will apply to a bunch of stuff so `mne` is probably the right place

**Pull request closed without merging**